include/stream-client/logger.hpp

```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+#include <chrono>
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+namespace stream_client {
+
+/**
+ * Base enum of log log levels
+ */
+enum class log_level {
+    error = 0,
+    warning,
+    info,
+};
+
+/**
+ * Base class for log events from internals of library
+ */
+class log_interface {
+public:
+    virtual ~log_interface() {}
+
+    /**
+     * Message from library for application.
+     *
+     * @param[in] level Level of message, application may use correct it according the target if necessary.
+     * @param[in] target Target of message is used for filter messages from different parts of library.
+     * @param[in] message The body and main reason of message.
+     * @param[in] error The error instance if exists.
+     *
+     * @note It can be from multiple threads, you should worry about sync and not lock this calls.
+     */
+    virtual void message(log_level level, const std::string& target, const std::string& message, const std::runtime_error* error = nullptr) const = 0;
+};
+
+/**
+ * Set logger instace for library
+ *
+ * @param[in] logger Logger interface for usage.
+ *
+ * @note You should do this once when starting the application and the lifetime of the variable till exit.
+ */
+void set_logger(const log_interface& logger);
```

Comment:
This is good, but it would be nice if we allow the user to pass just a callback, like a lambda or whatever, and in such case construct a logger instance that would just call this callback. Something like `void set_logger(std::function<void(log_level level, const std::string& location, const std::string& message)> callback);`. This allows a much simpler integration when the user is not supposed to implement an adapter class for `log_interface` and just pass a callback.